### PR TITLE
[MRG+1] doc for creating custom cache storage backend.

### LINKED ISSUE #3683: Document how to implement custom HTTP cache storage backed & cache policy.
https://docs.scrapy.org/en/latest/topics/downloader-middleware.html#module-scrapy.downloadermiddlewares.httpcache does let readers know they could implement custom HTTP cache storage backend and cache policy, but provides no information about the API.  I would like to work on documenting it. Should I go ahead?

### PR DESCRIPTION
**Pull request by anubhavp28**
Fixes #3683

### CODE REVIEW

File: docs/topics/downloader-middleware.rst [hunk 1]
```diff
@@ -496,6 +496,59 @@ In order to use this storage backend:
...
+You can implement custom cache storage backend by creating a Python class that
+defines the methods described below.
+
+.. module:: scrapy.extensions.httpcache
+
+.. class:: CacheStorage
+
+    .. method:: open_spider(spider)
+
+      This method gets called after a spider has been opened for crawling.
```
Comment (Gallaecio) [inline docs/topics/downloader-middleware.rst]: You should probably mention that this method handles the `open_spider` signal, and link to the documentation of that signal using the special `:signal:` syntax, e.g. ``` :signal:`spider_opened` ```. Same for `close_spider` below.

File: docs/topics/downloader-middleware.rst [hunk 2]
```diff
@@ -496,6 +496,59 @@ In order to use this storage backend:
...
+      :type request: :class:`~scrapy.http.Request` object
+
+    .. method:: store_response(spider, request, response)
+
+      Stores the given response in the cache.
+
+      :param spider: the spider for which the response is intended
+      :type spider: :class:`~scrapy.spiders.Spider` object
+
+      :param request: corresponding request the spider generated
```
Comment (Gallaecio) [inline docs/topics/downloader-middleware.rst]:
```suggestion
      :param request: the corresponding request the spider generated
```

File: docs/topics/downloader-middleware.rst [hunk 3]
```diff
@@ -496,6 +496,59 @@ In order to use this storage backend:
...
+      :type request: :class:`~scrapy.http.Request` object
+
+      :param response: the response to store in the cache
+      :type response: :class:`~scrapy.http.Response` object
+
+In order to use your storage backend, set:
+
+* :setting:`HTTPCACHE_STORAGE` to path of your custom storage class.
```
Comment (Gallaecio) [inline docs/topics/downloader-middleware.rst]:
```suggestion
* :setting:`HTTPCACHE_STORAGE` to the Python import path of your custom storage class.
```

**Comment by anubhavp28** [comment]
Hey @Gallaecio, sorry for replying so late. I have made the changes suggested by you. Can you review this again? Can we delay merging this? I am working on the documentation for creating custom cache policy. I think, we should include that as well inside this pull request.

**Comment by Gallaecio** [comment]
@anubhavp28 Actually, I think that it is better to keep pull requests separate when possible. The smaller a pull request is, the easier it is to review it.

File: docs/topics/downloader-middleware.rst [hunk 4]
```diff
@@ -349,7 +349,7 @@ HttpCacheMiddleware
         * :ref:`httpcache-storage-leveldb`
 
     You can change the HTTP cache storage backend with the :setting:`HTTPCACHE_STORAGE`
-    setting. Or you can also implement your own storage backend.
+    setting. Or you can also :ref:`implement your own storage backend. <httpcache-storage-custom>`
```
Comment (Gallaecio) [inline docs/topics/downloader-middleware.rst]:
```suggestion
    setting. Or you can also :ref:`implement your own storage backend <httpcache-storage-custom>`.
```

**Comment by kmike** [comment]
Thanks @anubhavp28 for improved docs, and @Gallaecio for the review!

**Pull request merged**